code/gfx/im_pipeline.cxx

```diff
@@ -0,0 +1,281 @@
+                      vs_blob->GetBufferPointer(),
+                      vs_blob->GetBufferSize(),
+                      &gD3d.xxc_pipeline->layout
+                      );
+  d3d_check_hresult_(hr);
+
+  d3d_safe_release_(err_blob);
+  d3d_safe_release_(vs_blob);
+  d3d_safe_release_(ps_blob);
+}
+
+void
+gfx_im_vertex(Vec2 position, u32 color) {
+  XXC_Pipeline &p = *gD3d.xxc_pipeline;
+
+  p.v[p.v_mark] = Vertex_XXC{position, color};
+  p.v_mark++;
+
+  if (p.v_mark == sizeof(p.v)/sizeof(*p.v)) {
+    gfx_im_flush();
+  }
+}
+
+void
+gfx_im_rect(Vec2 position, Vec2 size, u32 color) {
+  Vec2 const p[6] = {
+    {position.x         , position.y},         /* top-left */
+    {position.x + size.x, position.y},         /* top-right */
+    {position.x + size.x, position.y - size.y},/* btm-right */
+    {position.x + size.x, position.y - size.y},/* btm-right */
+    {position.x         , position.y - size.y},/* btm-left */
+    {position.x         , position.y}          /* top-left */
+  };
+
+  for (int i = 0; i < 6; i++) {
+    gfx_im_vertex(p[i], color);
+  }
+}
+
+void
+gfx_im_flush() {
+  ::HRESULT hr;
+  ::ID3D11DeviceContext &device_ctx = *gD3d.device_context;
+  XXC_Pipeline          &pipeline   = *gD3d.xxc_pipeline;
+  ::UINT const vertex_stride = sizeof(Vertex_XXC);
+  ::UINT const vertex_offset = 0;
+  ::UINT const vertex_count  = (::UINT)pipeline.v_mark;
+  
+  pipeline.v_mark = 0;
+
+  // Copy new vertices to VRAM
+  {
+    ::D3D11_MAPPED_SUBRESOURCE vbo_data = {0};
+    hr = device_ctx.Map(pipeline.vbo, 0, D3D11_MAP_WRITE_DISCARD, 0, &vbo_data);
+    d3d_check_hresult_(hr);
+
+    mem_copy_(vbo_data.pData, pipeline.v, sizeof(Vertex_XXC)*vertex_count);
+    device_ctx.Unmap(pipeline.vbo, 0);
+  }
+
+  // Set the constants
+  {
+    ::D3D11_MAPPED_SUBRESOURCE consts_data = {0};
+    hr = device_ctx.Map(pipeline.consts, 0, D3D11_MAP_WRITE_DISCARD, 0, &consts_data);
+    d3d_check_hresult_(hr);
+
+    mem_copy_(consts_data.pData, &pipeline.constants, sizeof(Constants));
+    device_ctx.Unmap(pipeline.consts, 0);
+  }
+
+  // Set the input assembler - what vertex buffer to draw, its layout, topology 
+
+  device_ctx.IASetPrimitiveTopology(D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
+  device_ctx.IASetInputLayout(pipeline.layout);
+  device_ctx.IASetVertexBuffers(0, 1, &pipeline.vbo, &vertex_stride, &vertex_offset);
+
+  // Set the pipeline
+  device_ctx.VSSetShader(pipeline.vs, NULL, 0);
+  device_ctx.VSSetConstantBuffers(0, 1, &pipeline.consts);
+
+  device_ctx.RSSetViewports( 1, &pipeline.viewport );
+  device_ctx.RSSetState(pipeline.rasterizer_state);
+
+  device_ctx.PSSetShader(pipeline.ps, NULL, 0);
+
+  // Draw
+  device_ctx.Draw(vertex_count, 0);
+}
+
+void
+gfx_im_create_vbo_or_panic() {
+  ::D3D11_BUFFER_DESC const vbo_desc = {
+    .ByteWidth      = sizeof(XXC_Pipeline::v),
+    .Usage          = D3D11_USAGE_DYNAMIC,
+    .BindFlags      = D3D11_BIND_VERTEX_BUFFER,
+    .CPUAccessFlags = D3D11_CPU_ACCESS_WRITE
+  };
```

Comment:
It is better to create **D3D11_USAGE_IMMUTABLE** resource and pass it an **D3D11_SUBRESOURCE_DATA** with initial data for vertex/index buffers to ensure they will end up in VRAM. This way you don't need any CPU access rights.